File: WikiCodeParser/Tags/WikiLinkTag.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;
using WikiCodeParser.Models;
using WikiCodeParser.Nodes;

namespace WikiCodeParser.Tags
{
    public class WikiLinkTag : LinkTag
    {
        public WikiLinkTag()
        {
            Token = null;
        }

        public override bool Matches(State state, string token)
        {
            var pt = state.PeekTo("]]");
            return pt?.Length > 1 && pt[1] == '[' && !pt.Contains("\n")
                   && Regex.IsMatch(pt.Substring(2), @"([^\]]*?)(?:\|([^\]]*?))?", RegexOptions.IgnoreCase);
        }

        public override INode Parse(Parser parser, State state, string scope)
        {
            var index = state.Index;

            if (state.Next() != '[' || state.Next() != '[')
            {
                state.Seek(index, true);
                return null;
            }

            var str = state.ScanTo("]]");

            if (state.Next() != ']' || state.Next() != ']')
            {
                state.Seek(index, true);
                return null;
            }

            var match = Regex.Match(str, @"^([^\]]+?)(?:\|([^\]]*?))?$", RegexOptions.IgnoreCase);
            if (!match.Success)
            {
                state.Seek(index, true);
                return null;
            }

            var page = System.Web.HttpUtility.HtmlDecode(match.Groups[1].Value);
            var text = match.Groups[2].Success ? match.Groups[2].Value : page;
            
            var url = System.Web.HttpUtility.HtmlAttributeEncode($"https://twhl.info/wiki/page/{WikiRevision.CreateSlug(page)}");
            var before = $"<a href=\"{url}\">";
            var after = "</a>";
            
            var content = new NodeCollection();
            content.Nodes.Add(new MetadataNode("WikiLink", page));
            content.Nodes.Add(new PlainTextNode(text));

            return new HtmlNode(before, content, after);
        }
    }
}

File: javascript/src/Tags/WikiLinkTag.ts
import { Parser } from '..';
import { HtmlHelper } from '../HtmlHelper';
import { WikiRevision } from '../Models/WikiRevision';
import { HtmlNode } from '../Nodes/HtmlNode';
import { INode } from '../Nodes/INode';
import { MetadataNode } from '../Nodes/MetadataNode';
import { NodeCollection } from '../Nodes/NodeCollection';
import { PlainTextNode } from '../Nodes/PlainTextNode';
import { ParseData } from '../ParseData';
import { State } from '../State';
import { TagParseContext } from '../TagParseContext';
import { Tag } from './Tag';

export class WikiLinkTag extends Tag {
    constructor() {
        super();
        this.Token = null;
    }

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    public override Matches(state: State, _token: string, _context: TagParseContext): boolean {
        const pt = state.PeekTo(']]');
        return pt?.length > 1 && pt[1] == '[' && !pt.includes('\n')
            && pt.substring(2).match(/([^\]]*?)(?:\|([^\]]*?))?/i) != null;
    }

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    public override Parse(_parser: Parser, _data: ParseData, state: State, _scope: string, _context: TagParseContext): INode {
        const index = state.Index;

        if (state.Next() != '[' || state.Next() != '[') {
            state.Seek(index, true);
            return null;
        }

        const str = state.ScanTo(']]');

        if (state.Next() != ']' || state.Next() != ']') {
            state.Seek(index, true);
            return null;
        }

        const match = str.match(/^([^\]]+?)(?:\|([^\]]*?))?$/i);
        if (!match) {
            state.Seek(index, true);
            return null;
        }

        let page = match[1];
        const text = match[2] ? match[2] : page;
        let hash = '';
        if (page.includes('#')) {
            const spl = page.split('#');
            page = spl[0];
            const anchor = spl.length > 1 ? spl.slice(1).join('#') : '';
            hash = '#' + anchor.replace(/[^\da-z?/:@\-._~!$&'()*+,;=]/ig, '_');
        }

        const url = HtmlHelper.AttributeEncode(`https://twhl.info/wiki/page/${WikiRevision.CreateSlug(page)}`) + hash;
        const before = `<a href="${url}">`;
        const after = '</a>';

        const content = new NodeCollection();
        content.Nodes.push(new MetadataNode('WikiLink', page));
        content.Nodes.push(new PlainTextNode(text));

        return new HtmlNode(before, content, after);
    }
}
